erpnext/accounts/doctype/payment_entry/payment_entry.py

```diff
@@ -1170,6 +1173,61 @@ def set_remarks(self):
 
 		self.set("remarks", "\n".join(remarks))
 
+	@frappe.requires_permission("Account", "read")
+	@frappe.requires_permission("Sales Order", "read")
+	@frappe.requires_permission("Payment Entry", "create")
+	def _from_sales_order(self, so):
```

Comment:
where are these methods used?

Reply:
This particular method can be used through their entry point(s) like so:

```python
pe = frappe.new_doc_from("Payment Entry", "Sales Order", "SO-001")
# or
so = frappe.get_doc("Sales Order", "SO-001")
pe = so.into("Payment Entry")
# or
pe = frappe.new_doc("Payment Entry")
pe.flags.some_flag = True
# with either
pe.from_doc("Sales Order", "SO-001")
# or with
so = frappe.get_doc("Sales Order", "SO-001")
pe.from_doc(so)
```

While I did originally consider to mangle these methods `__*` so that they would be hard to directly invoke outside of the instance, that approach wasn't particularly forward-compatible with inheritance/overrides so I droppped it again and reverted to content with only a conventionalized `_*` private method marker.

Reply:
The decorators (`@frappe.requires_permission`) whitelist fine grained individual access rights at the doctype level. Meaning: not at the record level, because we can't sensibly declare record level acl in generic code.

This does two things:

- It marginally improves on the principle of least privilege vs a blanket `self.flags.ignore_permissions`, but the effect of this is somewhat limited because we only gain granularity in the permission dimension ("read" vs "write"), doctype granularity was already implemented at the code leaves before.
- But more importantly, it declares a security context for conversion methods which can be reasoned about as a enclosed and well, specified declared context so that a reader of the code can rely on that declaration (provided the implementation is correct and not buggy).

Reply:
Additional note, which triggered this PR series:

Unlike combinations of `make_*` methods, this now works with server scripts, such as the post payment processing methods of: https://github.com/frappe/frappe/pull/25952

Coincidentally, this is also the reason why this PR is heavy on the _Payment Entry_ doctype which is most often involved in these particular hooks.